pkg/utils/secrets/manager/get_or_generate.go

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manager
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	secretutils "github.com/gardener/gardener/pkg/utils/secrets"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (m *manager) GetOrGenerate(ctx context.Context, config secretutils.ConfigInterface, opts ...GetOrGenerateOption) (*corev1.Secret, error) {
+	options := &GetOrGenerateOptions{}
+	if err := options.ApplyOptions(m, config, opts); err != nil {
+		return nil, err
+	}
+
+	var bundleFor *string
+	if options.isBundleSecret {
+		bundleFor = pointer.String(strings.TrimSuffix(config.GetName(), nameSuffixBundle))
+	}
+
+	objectMeta, err := ObjectMeta(m.namespace, config, m.lastRotationStartTimes[config.GetName()], options.signingCAChecksum, &options.Persist, bundleFor)
+	if err != nil {
+		return nil, err
+	}
+
+	secret := &corev1.Secret{}
+	if err := m.client.Get(ctx, kutil.Key(objectMeta.Namespace, objectMeta.Name), secret); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+
+		secret, err = m.generateAndCreate(ctx, config, objectMeta)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !options.isBundleSecret {
+		if err := m.addToStore(config.GetName(), secret, current); err != nil {
+			return nil, err
+		}
+
+		if !options.IgnoreOldSecrets && options.RotationStrategy == KeepOld {
+			if err := m.storeOldSecrets(ctx, config.GetName(), secret.Name); err != nil {
+				return nil, err
+			}
+		}
+
+		if err := m.getOrGenerateBundleSecret(ctx, config); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := m.reconcileSecret(ctx, secret, objectMeta.Labels); err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+func (m *manager) generateAndCreate(ctx context.Context, config secretutils.ConfigInterface, objectMeta metav1.ObjectMeta) (*corev1.Secret, error) {
+	data, err := config.Generate()
+	if err != nil {
+		return nil, err
+	}
+
+	secret := Secret(objectMeta, data.SecretData())
+	if err := m.client.Create(ctx, secret); kutil.IgnoreAlreadyExists(err) != nil {
```

Comment:
This is mainly to deal with stale caches - if the secret already exists in the system with the same name then we would not like to ovewrite its data it (would not be possible anyways as it's immutable).
Besides, it also helps us for migration purposes since for existing shoot clusters the secrets for the CAs still exist with the same name and we don't want to delete them.

But yes, I agree to the point making an additional `Get` call to not write invalid data to the store. Good catch, thanks!